src/HPLTokenHandler.mo

```diff
@@ -0,0 +1,331 @@
+
+  public type Info = {
+    var credit : Nat;
+    virtualAccountId : HPL.VirtualAccountId;
+  };
+
+  public type StableInfo = {
+    credit : Nat;
+    virtualAccountId : HPL.VirtualAccountId;
+  };
+
+  public type JournalRecord = (Time.Time, Principal, {
+    #credited: Nat;
+    #debited: Nat;
+    #error: Any;
+    #newDeposit: Nat;
+    #openAccountError: { #UnknownPrincipal; #UnknownSubaccount; #MismatchInAsset; #NoSpaceForAccount };
+    #withdraw: { to: (Principal, HPL.AccountReference); amount: Nat };
+  });
+
+  public type WithdrawError = { 
+    #TooLargeFtQuantity;
+    #TooLargeSubaccountId;
+    #TooLargeVirtualAccountId;
+    #UnknownPrincipal;
+    #UnknownSubaccount;
+    #UnknownVirtualAccount;
+    #DeletedVirtualAccount;
+    #MismatchInAsset;
+    #MismatchInRemotePrincipal; 
+    #CallHPLError;
+  };
+
+  public type StableData = (
+    [(Principal, StableInfo)],        // map
+    Nat,                              // consolidatedFunds
+    (Nat, Nat),                       // total debited/credited
+    ([var ?JournalRecord], Nat, Int)  // journal
+  );
+
+  public class TokenHandler(
+    hplLedgerPrincipal : Principal,
+    assetId: HPL.AssetId, 
+    backingSubaccountId: HPL.SubaccountId,
+    ownPrincipal : Principal,
+    journalSize: Nat,
+  ) {
+    let hpl = actor (Principal.toText(hplLedgerPrincipal)) : HPL.Ledger;
+
+    /// if some unexpected error happened, this flag turns true and handler stops doing anything until recreated
+    var isFrozen_ : Bool = false;
+    func freezeTokenHandler(errorText: Text): () {
+      isFrozen_ := true;
+      journal.push((Time.now(), ownPrincipal, #error(errorText)));
+    };
+
+    var journal : CircularBuffer.CircularBuffer<JournalRecord> = CircularBuffer.CircularBuffer<JournalRecord>(journalSize);
+    var consolidatedFunds_ : Nat = 0;
+    var map : RBTree.RBTree<Principal, Info> = RBTree.RBTree<Principal, Info>(Principal.compare);
+
+    var totalDebited : Nat = 0;
+    var totalCredited : Nat = 0;
+
+    /// Returns reference to registered virtual account for P.
+    /// If not registered yet, registers it automatically
+    public func getAccountReferenceFor(p : Principal) : async* (Principal, HPL.VirtualAccountId) {
+      let virtualAccountId : HPL.VirtualAccountId = switch (map.get(p)) {
+        case (?info) info.virtualAccountId;
+        case (null) {
+          let registerResult = await hpl.openVirtualAccount({ asset = #ft(assetId, 0); backingSubaccountId = backingSubaccountId; remotePrincipal = p });
+          switch (registerResult) {
+            case (#ok vid) {
+              map.put(p, {
+                virtualAccountId = vid;
+                var credit = 0;
+              });
+              vid;
+            };
+            case (#err error) {
+              let message = "Opening virtual account problem";
+              journal.push((Time.now(), p, #error(message, error)));
+              freezeTokenHandler(message);
```

Comment:
We should freeze only if there is an _internal_ error. In other words, a bug.

If a virtual account cannot be opened then we cannot serve the new principal. But the TokenHandler is still operational for everyone else.

Reply:
removed logic to freeze handler in this case